File: leetcode/problems/problem-specific/102.binary-tree-level-order-traversal/102.binary-tree-level-order-traversal.go
package main

/*
 * @lc app=leetcode id=102 lang=golang
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start

// * Iterative BFS Solution | O(n) Time | O(n) Space
// * n -> Number of nodes in a tree

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	res, queue := [][]int{}, []*TreeNode{root}
	for len(queue) > 0 {
		curQueueSize, curLevelElements := len(queue), []int{}
		for i := 0; i < curQueueSize; i++ {
			curNode := queue[0]
			queue = queue[1:]
			curLevelElements = append(curLevelElements, curNode.Val)
			if curNode.Left != nil {
				queue = append(queue, curNode.Left)
			}
			if curNode.Right != nil {
				queue = append(queue, curNode.Right)
			}
		}

		res = append(res, curLevelElements)
	}

	return res
}

// @lc code=end

File: leetcode/problems/problem-specific/102.binary-tree-level-order-traversal/102.binary-tree-level-order-traversal.py
#
# @lc app=leetcode id=102 lang=python3
#
# [102] Binary Tree Level Order Traversal
#

# @lc code=start

# * Iterative BFS Solution | O(n) Time | O(n) Space
# * n -> The number of nodes in the tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        if not root:
            return result

        # * BFS Traversal
        queue = collections.deque([root])
        while queue:
            level_order_values = []
            for _ in range(len(queue)):
                cur_node = queue.popleft()
                level_order_values.append(cur_node.val)
                if cur_node.left:
                    queue.append(cur_node.left)
                if cur_node.right:
                    queue.append(cur_node.right)

            result.append(level_order_values)

        return result

# @lc code=end
